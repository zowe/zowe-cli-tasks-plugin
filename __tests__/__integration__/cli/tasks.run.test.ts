/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { TestEnvironment } from "../../__src__/environment/TestEnvironment";
import { ITestEnvironment } from "../../__src__/environment/doc/response/ITestEnvironment";
import { runCliScript } from "../../__src__/TestUtils";
import * as fs from "fs";

import * as path from "path";
const dirs = (p: any) => fs.readdirSync(p).filter((f) => fs.statSync(path.join(p, f)).isDirectory());

let TEST_ENVIRONMENT: ITestEnvironment;
describe("run tasks command", () => {
    beforeAll(async () => {
        TEST_ENVIRONMENT = await TestEnvironment.setUp({
            testName: "run_tasks_command",
            installPlugin: true,
            skipProperties: true
        });
    });

    afterAll(async () => {
        await TestEnvironment.cleanUp(TEST_ENVIRONMENT);
    });

    it("should be able to substitute values from ENV and global in desc and scripts", () => {
        const yml = path.join(TEST_ENVIRONMENT.workingDir, "zowe-tasks-substitution.yml");
        fs.copyFileSync(path.join(__dirname, "__yaml__", "zowe-tasks-substitution.yml"), yml);
        const output = runCliScript(path.join(__dirname, "__scripts__", "zowe-tasks-substitution.sh"),
            TEST_ENVIRONMENT, [yml]);
        expect(output.stdout.toString()).toMatchSnapshot();
        expect(output.status).toBe(0);
    });

    it("should be able to run a CLI command", () => {
        const yml = path.join(TEST_ENVIRONMENT.workingDir, "zowe-tasks-run-cli.yml");
        fs.copyFileSync(path.join(__dirname, "__yaml__", "zowe-tasks-run-cli.yml"), yml);
        const output = runCliScript(path.join(__dirname, "__scripts__", "zowe-tasks-run-cli.sh"),
            TEST_ENVIRONMENT, [yml]);
        expect(output.stdout.toString()).toMatchSnapshot();
        expect(output.status).toBe(0);
        const dateTimeDir = dirs(path.join(TEST_ENVIRONMENT.workingDir, "zowe-tasks-out-run-cli"));
        const outputLogPath = path.join(TEST_ENVIRONMENT.workingDir,
            "zowe-tasks-out-run-cli",
            dateTimeDir[0],
            "task_sequence_1_run-zowe-cli",
            "action_sequence_2_runCommand",
            "3_runCommand.action.output.txt");
        const logContents = fs.readFileSync(outputLogPath).toString();
        const json = JSON.parse(logContents);
        expect(json.success).toBe(true);
        expect(json.stdout).toContain("\nInstalled plugins: \n\n -- pluginName: @zowe/tasks-for-zowe-cli");
    });

    it("should give stdout and stderr if an exec command fails", () => {
        fs.copyFileSync(path.join(__dirname, "__yaml__", "zowe-tasks.yml"),
            path.join(TEST_ENVIRONMENT.workingDir, "zowe-tasks.yml"));
        const output = runCliScript(path.join(__dirname, "__scripts__", "run_tasks.sh"), TEST_ENVIRONMENT,
            [path.join(__dirname, "__scripts__", "stdout_and_stderr.sh")]);
        expect(output.stderr.toString()).toContain("Failed to exec command:");
        expect(output.stderr.toString()).toContain("Hello stderr!");
        expect(output.stderr.toString()).toContain("Command Stdout:");
        expect(output.stderr.toString()).toContain("Hello stdout!");
        expect(output.stderr.toString()).toContain("Command failed with exit code \"1\"");
        expect(output.stderr.toString()).toContain("The action in progress was \"execCommand\".");
        expect(output.status).toBe(1);
    });
});
