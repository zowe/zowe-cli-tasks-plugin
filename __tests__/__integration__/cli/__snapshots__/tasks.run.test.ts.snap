// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run tasks command should be able to repeat an action over an iterable with forEach 1`] = `
"
Task - repeat-foreach - \\"repeat the action until the validators mark it valid\\"
aardvark
   ✔   Action (printUbuntuAnimals) \\"print animal for Ubuntu release: artful\\"
beaver
   ✔   Action (printUbuntuAnimals) \\"print animal for Ubuntu release: bionic\\"
cuttlefish
   ✔   Action (printUbuntuAnimals) \\"print animal for Ubuntu release: cosmic\\"
dingo
   ✔   Action (printUbuntuAnimals) \\"print animal for Ubuntu release: disco\\"
ermine
   ✔   Action (printUbuntuAnimals) \\"print animal for Ubuntu release: eoan\\"
   ✔   Action (numberUbuntuReleases) \\"calculate major number for each Ubuntu release\\"
   ✔   Action (checkForEachExtracted) \\"check that variables get resolved in repeat.forEach (artful)\\"
   ✔   Action (checkForEachExtracted) \\"check that variables get resolved in repeat.forEach (bionic)\\"
   ✔   Action (checkForEachExtracted) \\"check that variables get resolved in repeat.forEach (cosmic)\\"
   ✔   Action (checkForEachExtracted) \\"check that variables get resolved in repeat.forEach (disco)\\"
   ✔   Action (checkForEachExtracted) \\"check that variables get resolved in repeat.forEach (eoan)\\"
"
`;

exports[`run tasks command should be able to repeat an action until the validators mark it valid 1`] = `
"
Task - repeat-until - \\"repeat the action until the validators mark it valid\\"
   ✔   Action (getFutureTime) \\"get a time in the future\\"
   ✔   Action (checkCurrentTime) \\"check if the current time is beyond the future time\\"
"
`;

exports[`run tasks command should be able to run a CLI command 1`] = `
"
Task - run-zowe-cli - \\"Run a Zowe CLI command\\"
   ✔   Action (runCommand) \\"Run a Zowe CLI Command\\"
"
`;

exports[`run tasks command should be able to substitute values from ENV and global in desc and scripts 1`] = `
"
Task - substitution - \\"var1: Var 1 Value var2: Var 2 Value var3: Var 3 Value\\"
   ✔   Action (Run some javascript) \\"Var 1 ValueVar 2 ValueVar 3 Value\\"
"
`;
